src/Argument.h

```diff
@@ -53,27 +68,14 @@ struct Argument {
      * Note that type.lanes should always be 1 here. */
     Type type;
 
-    /** If this is a scalar parameter, then these are its default, min, max values.
-     * By default, they are left unset, implying "no default, no min, no max". */
-    Expr def, min, max;
+    /* The estimates (if any) and default/min/max values (if any) for this Argument. */
+    ArgumentEstimates argument_estimates;
 
-    Argument() : kind(InputScalar), dimensions(0) {}
+    Argument() = default;
     Argument(const std::string &_name, Kind _kind, const Type &_type, int _dimensions,
-                Expr _def = Expr(),
-                Expr _min = Expr(),
-                Expr _max = Expr()) :
-        name(_name), kind(_kind), dimensions((uint8_t) _dimensions), type(_type), def(_def), min(_min), max(_max) {
-        internal_assert(_dimensions >= 0 && _dimensions <= 255);
-        user_assert(!(is_scalar() && dimensions != 0))
-            << "Scalar Arguments must specify dimensions of 0";
-        user_assert(!(is_buffer() && def.defined()))
-            << "Scalar default must not be defined for Buffer Arguments";
-        user_assert(!(is_buffer() && min.defined()))
-            << "Scalar min must not be defined for Buffer Arguments";
-        user_assert(!(is_buffer() && max.defined()))
-            << "Scalar max must not be defined for Buffer Arguments";
-    }
+             const ArgumentEstimates &argument_estimates);
 
+    // TODO: this should really be marked explicit
```

Comment:
... but then you wouldn't be able to (without a cast) put a Buffer in an argument list, to indicate that it shouldn't be baked into the object file, but instead received as an argument at runtime.

Reply:
true. it's just my knee-jerk assumption that non-explicit ctors should be regarded with suspicion. I'll update the comment.